### PR TITLE
Bugfix: Fix DirectTreeOptimizer for clips with partial keyframes

### PR DESCRIPTION
**Pull request by MorghusDragon**
When animating for example the rotation of an object, it's possible to skip the "end" keyframe on some of the properties and Unity uses the value of the first keyframe for the rest of the curve.

For the Direct Tree Optimizer, the end clip is calculated from the curves, however, it does currently not take into account those clips that only have a "start" keyframe and no "end" keyframe.

This change looks at the curves and if there's only one keyframe, it copies the value of the "start" clip to the "end" clip.

Example clip that would have been broken (notice the missing keyframes at the end of the clip):
image

Example clip that works:
image

```
The changes made in this contribution are free and unencumbered software
released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>
```

### CODE REVIEW

File: com.vrcfury.vrcfury/Editor/VF/Service/ClipBuilderService.cs [hunk 1]
```diff
@@ -171,10 +171,16 @@ public class ClipBuilderService {
                     foreach (var key in curve.FloatCurve.keys) {
                         (key.time == 0 ? startClip : endClip).SetConstant(binding, key.value);
                     }
+                    if (curve.FloatCurve.keys.Length == 1) {
+                        endClip.SetConstant(binding, curve.FloatCurve.keys[0].value);
```
Comment (SenkyDragon) [inline com.vrcfury.vrcfury/Editor/VF/Service/ClipBuilderService.cs:175]: Is it possible for a binding to have 0 frames? Would it make more sense to change the foreach to
```
var first = false;
foreach (var key in curve.FloatCurve.keys) {
   if (first) {
     startClip.SetConstant(binding, key.value);
     first = false;
   }
   endClip.SetConstant(binding, key.value);
}
```

Comment (SenkyDragon) [reply com.vrcfury.vrcfury/Editor/VF/Service/ClipBuilderService.cs:175]: I guess I'm also not sure what happens if the clip doesn't have a keyframe at time = 0. I'm also not sure if the frames are in time order 🤔

Comment (MorghusDragon) [reply com.vrcfury.vrcfury/Editor/VF/Service/ClipBuilderService.cs]: That makes sense, I updated it with your snippet, that will cover more cases.

Comment (SenkyDragon) [reply com.vrcfury.vrcfury/Editor/VF/Service/ClipBuilderService.cs]: seems like if there isn't a keyframe at time=0, time=0 inherits the first frame.

**Pull request merged**